### DISCUSSION TITLE
Entity refresh delay.

### DISCUSSION THREAD

**Discussion by rofo69** [jbergler/hass-ttlock#146] [Q&A]
HI all,

I have a raykube F7 lock thats working perfectly with this integration.
If I use a service call to add a passcode or change passage mode, its reflects almost instantly in the ttlock app, and the lock itself.

However, the update of the entity in HA can take 30 seconds or more to catch up with the current state of the lock. I'm guessing this is just the delay by the ttlock servers calling the webhook ? Interesting tho that the lock and app update almost instantly.

A bit more testing, and it only seems to be the passage mode sensor thats delayed. Locking or unlocking the door updates the entities almost instantly. And sometimes passage mode never seems to update.. last test its been minutes and still nothing.

Final update.. .it seems to work reliably and quickly if I make an HA update entities call afterwards, so I guess I'll just bake that into my automation and ESPhome code.

**Comment by jbergler** [comment]
Hiya - the app uses a mix of bluetooth and their cloud to achieve faster updates.
Given this integration doesn't use bluetooth (we'd have to re-implement their entire undocumented protocol, since we can't use the sdk) it's reliant on their cloud. 

I also choose not to use polling in the interest of saving battery life on the lock.

If you turn on debug logging for the integration (or for webhooks) you should be able to get some accurate timestamps to see if the delay is in their cloud or in home assistant, but I suspect it will be their cloud.

**Reply by rofo69** [reply]
Thanks, I'll give it a try.
Its interesting _**via HA**_ the outbound lock/unlock and code setting, deleting, and passage mode, all seem to be actioned within a second or two on the lock, the only delay is the message coming back, unless I force an update entity command in HA.

However, setting things like passage mode _**in the ttlock app**_ seems to be very very slow in updating HA, if ever, and it doesn't matter if force an update entities call or not.

**Reply by jbergler** [reply]
A lot of those changes don't have corresponding webhooks, so theres no way for us to know something changed in HA.

**Reply by rofo69** [reply]
I was just realising that as I was testing.  Other than remote unlock that needs the gateway, I'm guessing the rest of the commands are done via bluetooth, so the app is talking directly to the lock and the lock isn't updating the cloud with its status. 
I wonder if this is also why I can't actually see the status of the lock in the ttlock app, at least I can't find it. I was assuming the main lock icon would change to unlocked when the lock is in that state, but it doesnt.

Ignore the lock status thing above, I was unaware you needed to press the little hexagon icon on the top right of the screen to see the status of the lock.

**Reply by jbergler** [reply]
The docs are not very good, but you can see how little the notification (Webhook) method actually tells us here: https://euopen.ttlock.com/document/doc?urlName=cloud%2FlockRecord%2FnotifyEn.html

**Reply by rofo69** [reply]
Thats not alot.
I've also noticed you the service call you implement for passage mode, allows you to pass an auto unlock boolean, so when passage mode is set, the door automatically unlocks. I've tried this and it doesn't not seem to work. I've looked in the ttlock app I can't find this option either when setting passage mode, so is it possible they've removed it ?

**Reply by rofo69** [reply]
EDIT: I see in the docs auto unlock is still supported, so that might have been my mistake. I was passing a boolean parameter when it wants an int. I see when I use dev tools its using bools, ah I see the code converts these:-

                passageMode=1 if config.enabled else 2,
                autoUnlock=1 if config.auto_unlock else 2,

So for me at least, the auto unlock feature as part of turning on passage mode doesn't appear to work.
